internal/runtimehandlerhooks/high_performance_hooks.go

```diff
@@ -356,6 +417,171 @@ func setCPUQuota(cpuMountPoint, parentDir string, c *oci.Container, enable bool)
+			latencyFile := fmt.Sprintf("%s/cpu%d/power/pm_qos_resume_latency_us", cpuDir, cpu)
+			latencyOrig, err := ioutil.ReadFile(latencyFile)
+			if err != nil {
+				return err
+			}
+
+			// save the current latency so we can restore it later
+			cpuPowerSaveDir := fmt.Sprintf("%s/cpu%d/power", cpuSaveDir, cpu)
+			err = os.MkdirAll(cpuPowerSaveDir, 0o750)
+			if err != nil {
+				return err
+			}
+			latencyFileOrig := fmt.Sprintf("%s/pm_qos_resume_latency_us", cpuPowerSaveDir)
+			err = ioutil.WriteFile(latencyFileOrig, latencyOrig, 0o644)
+			if err != nil {
+				return err
+			}
+
+			// update the pm_qos_resume_latency_us
+			err = ioutil.WriteFile(latencyFile, []byte(latency), 0o644)
+			if err != nil {
+				return err
+			}
+		} else {
+			// retrieve the original latency
+			latencyFileOrig := fmt.Sprintf("%s/cpu%d/power/pm_qos_resume_latency_us", cpuSaveDir, cpu)
+			latencyOrig, err := ioutil.ReadFile(latencyFileOrig)
+			if err != nil {
+				if errors.Is(err, os.ErrNotExist) {
+					// the latency may have already been restored by a previous invocation of the hook
+					return nil
+				}
+				return err
+			}
+
+			// restore the original latency
+			latencyFile := fmt.Sprintf("%s/cpu%d/power/pm_qos_resume_latency_us", cpuDir, cpu)
+			err = ioutil.WriteFile(latencyFile, latencyOrig, 0o644)
+			if err != nil {
+				return err
+			}
+
+			// remove the saved latency
+			err = os.Remove(latencyFileOrig)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// setCPUFreqGovernor sets the scaling_governor for a cpu and stores the original
+// value so it can be restored later. If the governor is an empty string, the original
+// scaling_governor value is restored.
+func setCPUFreqGovernor(c *oci.Container, governor, cpuDir, cpuSaveDir string) error {
+	lspec := c.Spec().Linux
+	if lspec == nil ||
+		lspec.Resources == nil ||
+		lspec.Resources.CPU == nil ||
+		lspec.Resources.CPU.Cpus == "" {
+		return errors.Errorf("find container %s CPUs", c.ID())
+	}
+
+	cpus, err := cpuset.Parse(lspec.Resources.CPU.Cpus)
+	if err != nil {
+		return err
+	}
+
+	for _, cpu := range cpus.ToSlice() {
+		if governor != "" {
+			// retrieve the current scaling governor
+			governorFile := fmt.Sprintf("%s/cpu%d/cpufreq/scaling_governor", cpuDir, cpu)
+			governorOrig, err := ioutil.ReadFile(governorFile)
+			if err != nil {
+				return err
+			}
+
+			// get available cpu scaling governors
+			availGovernorFile := fmt.Sprintf("%s/cpu%d/cpufreq/scaling_available_governors", cpuDir, cpu)
+			availGovernors, err := ioutil.ReadFile(availGovernorFile)
+			if err != nil {
+				return err
+			}
+
+			// is the scaling governor supported?
+			governorFound := false
+			for _, availableGovernor := range strings.Fields(string(availGovernors)) {
+				if availableGovernor == governor {
+					governorFound = true
+					break
+				}
+			}
+			if !governorFound {
+				return errors.Errorf("governor %s not available for cpu %d", governor, cpu)
+			}
```

Comment:
I think it is safe to assume that the list of available governors is the same for all CPUs, so it should only be checked once.

Yet better, you do not want to check it at all -- since writing of an invalid string to `scaling_governor` should result in a particular error (`unix.EINVAL`).

Reply:
So looking at the kernel code, it looks like there is only a single set of governors (stored in the cpufreq_governor_list global):
https://github.com/torvalds/linux/blob/268db333b561c77dee3feb6831806412293b4a7e/drivers/cpufreq/cpufreq.c#L2441

However, I think the way I've written this is more "correct" since the scaling_available_governors is exposed on a per cpu basis. My existing code is more future proof if the kernel implementation was ever to change. Is your concern related to the extra sysfs file access required for each cpu when a container is created (and has the new annotation)? If this is something where realtime is critical then I can make the change, but I just wanted to confirm we really need to do this.